app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistFragment.java

```diff
@@ -428,32 +429,33 @@ public void removeWatchedStreams(final boolean removePartiallyWatched) {
                             final var playlistItem = playlist.get(i);
                             final var streamStateEntity = streamStates.get(i);
 
-                            final int indexInHistory = Collections.binarySearch(historyStreamIds,
-                                    playlistItem.getStreamId());
                             final long duration = playlistItem.toStreamInfoItem().getDuration();
 
-                            if (indexInHistory < 0 || (streamStateEntity != null
-                                    && !streamStateEntity.isFinished(duration))) {
-                                notWatchedItems.add(playlistItem);
-                            } else if (!thumbnailVideoRemoved
-                                    && playlistManager.getPlaylistThumbnail(playlistId)
-                                    .equals(playlistItem.getStreamEntity().getThumbnailUrl())) {
-                                thumbnailVideoRemoved = true;
+                            if (streamStateEntity != null
+                                    && streamStateEntity.isFinished(duration)) {
+                                watchedItems.add(playlistItem);
```

Comment:
I think the point of this PR is just to change the conditions here. The previous conditions were flawed because a video present in history but without a stream state entity would be considered as fully watched therefore removed. But why would "having no stream state entity" mean "video is fully watched"? Therefore your condition looks good, although I would keep the `indexInHistory` check.

The previous conditions should hence be changed from
```java
indexInHistory < 0 || (streamStateEntity != null && !streamStateEntity.isFinished(duration))
```
to
```java
indexInHistory < 0 || streamStateEntity == null || !streamStateEntity.isFinished(duration)
```
which means "keep the video if it's not in history OR it has no stream state entity OR the stream is not finished", or in other words "remove the video if it's in history AND it has a stream state entity AND the stream is finished". This condition does make logical sense in my opinion.